cmake: add app dts root as search location for overlay files

pabigot opened the pull request:
Putting overlay files in the test/sample root clutters the file system.  Allow them to be in a dts subdirectory.

pabigot committed:
cmake: add app boards subdir as search location for board overlay files

Putting overlay files in the test/sample root clutters the file
system.  Allow them to be in the boards subdirectory alongside
board.conf files.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

SebastianBoe reviewed (changes requested):
This sounds reasonable and follows convention (app/boards/<board>.conf), but needs documentation.

pabigot commented:
@SebastianBoe OK, I'll do that.

Should this PR also include a move of all existing overlays down into a `dts` directory?  There's only a couple dozen each in `samples` and `tests`.  That would also permit deprecating their appearance in the application root.

SebastianBoe commented:
@pabigot : No, we can and should support both schemes. We don't want to break user application's for something trivial like this.

galak commented:
> @SebastianBoe OK, I'll do that.
> 
> Should this PR also include a move of all existing overlays down into a `dts` directory? There's only a couple dozen each in `samples` and `tests`. That would also permit deprecating their appearance in the application root.

I think it would be nice to move this in our `samples` & `tests` so we are consistent.  I agree with @SebastianBoe that i'd leave support for both locations.

doc/application/index.rst under 'Device Tree Overlays' seems like the proper place to update the docs.

pabigot commented:
@SebastianBoe @galak I agree about keeping both locations allowed, and with @galak about moving the existing. They're noisy.

A question though: We put `board_name.conf` into `boards`.  The overlay files being addressed here are all `board_name.overlay`.  They conceptually do something like "act as if this content was present in the board standard devicetree".  

There's likely to be some relation between content of the devicetree overlay and Kconfig options so having them co-located makes sense.  So I now would propose having the search path be:
```
if(DTC_OVERLAY_FILE)
  # DTC_OVERLAY_FILE has either been specified on the cmake CLI or is already
  # in the CMakeCache.txt. This has precedence over the environment
  # variable DTC_OVERLAY_FILE
elseif(DEFINED ENV{DTC_OVERLAY_FILE})
  set(DTC_OVERLAY_FILE $ENV{DTC_OVERLAY_FILE})
elseif(EXISTS          ${APPLICATION_SOURCE_DIR}/boards/${BOARD}.overlay)  # new
  set(DTC_OVERLAY_FILE ${APPLICATION_SOURCE_DIR}/boards/${BOARD}.overlay)
elseif(EXISTS          ${APPLICATION_SOURCE_DIR}/dts/${BOARD}.overlay)     # new
  set(DTC_OVERLAY_FILE ${APPLICATION_SOURCE_DIR}/dts/${BOARD}.overlay)
elseif(EXISTS          ${APPLICATION_SOURCE_DIR}/${BOARD}.overlay)
  set(DTC_OVERLAY_FILE ${APPLICATION_SOURCE_DIR}/${BOARD}.overlay)
elseif(EXISTS          ${APPLICATION_SOURCE_DIR}/app.overlay)
  set(DTC_OVERLAY_FILE ${APPLICATION_SOURCE_DIR}/app.overlay)
endif()
```

Do either of you object?

If not I'll move existing overlays into existing `boards` where available and to `dts` otherwise, in a separate commit.

pabigot committed:
samples: move board overlay files into boards directory

Application board.overlay files tend to be paired with
boards/board.conf files that extend the functionality of a board.
Move the overlay files to the same location as the config files that
they work with.

A few overlay files that are paired with a prj_board.conf file in the
application root directory are left in place.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

SebastianBoe commented:
@pabigot : Using "boards/" instead of "dts/" to get files out of the root app directory sounds reasonable to me.

galak commented:
Seems like this might need a rebase.

pabigot commented on doc/application/index.rst:
```diff
@@ -1535,17 +1535,23 @@ the C preprocessor on a file which includes the following:
-
-- If the file :file:`BOARD.overlay` exists in your application directory,
-  where ``BOARD`` is the BOARD value set earlier, it will be used.
+1. Any value given to :makevar:`DTC_OVERLAY_FILE` in your application
```
@dbkinder This was modeled after existing similar content for locating a configuration file in `Kconfig Configuration` earlier in the file.  Do you want me to change to this way only here, or to revise the other description as well as this?

pabigot replied:
In answer to your question: in both cases the first source that is present is used; the remainder are ignored.

dbkinder replied:
Understood about the first source wins, but in your text it says CMakeLists.txt, command line, and CMake cache all in #1, so I wanted to clarify that.

Looking back at the BOARD value description above, I see that the cache actually has the highest precedence, then the command line, then an environment variable, and finally the CMakeLists.txt. Is that true for this DTC_OVERLAY_FILE too (but we've deprecated the environment variable now), plus a couple more places to look.  

I was tripped up by the complicated sentences and parentheses here and thought a nice compact ordered list would clarify things. I'm OK with just making this suggested change in this section, the others don't seem as complicated.

pabigot replied:
@dbkinder This really opened a can of worms.  In the existing documentation I find four places where the precedence of build variables is specified.  None of them is the same as any other.

I believe in essence option #&#8203;1 expresses "from CMake", which can be any of the three listed sources, of which at least two are probably only consulted when the build area is initialized and between those two I don't know which takes precedence because cmake doesn't document precedence clearly anywhere I can find.

I've switched to your recommended text.

pabigot commented:
@dbkinder your requested text has been used, please revisit.
